### PR TITLE
TMU module

### PR DESCRIPTION
**Pull request by 3dawy**

### CODE REVIEW

File: Dynamic Design/SERVICE/TMU/TMU.c [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+  }str_FuncStatus_t;
+
+  /* this struct is used to determine the characteristics of each task */
+  typedef struct str_TMU_Buffer_t
+  {
+     
+	  uint8_t u8_Duty;                  /* whether the task is used or not */
+	  uint16_t u8_Delay_Time;           /* the wanted delay time */
+	  uint8_t u8_Repeat;                /* whether the task is repeated or one shot */
+	  void(* Ptr_Consumer)(void);       /* pointer to the consumer function */
```
Comment (hossamadel) [inline Dynamic Design/SERVICE/TMU/TMU.c]: Better to define it as typedef

File: Dynamic Design/SERVICE/TMU/TMU.c [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+	  uint8_t u8_Pre_Flag;              /* count the time for consumer function until reach it's periodicity  */
+	  uint8_t u8_Task_ID;               /* determine the Id for the consumer task */
+	  
+  }str_TMU_Buffer_t;
```
Comment (hossamadel) [inline Dynamic Design/SERVICE/TMU/TMU.c]: Naming convention ... recommended to be str_TmuBufferType for example

File: Dynamic Design/SERVICE/TMU/TMU.c [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+
+
+
+/********************************************************/
+/************ prototype for local functions ************/
+/******************************************************/
+	
+static ERROR_STATUS Is_Buffer_Empty(void);
+static ERROR_STATUS Is_Buffer_Full(void);
+static void Timer_CB (void);
```
Comment (hossamadel) [inline Dynamic Design/SERVICE/TMU/TMU.c]: Rename it to be timer_cbk

File: Dynamic Design/SERVICE/TMU/TMU.c [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+		else
+		{
+			u8_Ret_Status=FALSE;
+		}
+		return u8_Ret_Status;
+	}
+
+/* this function is called when timer over flow occur, this function is used to set a specific flag */
+ void Timer_CB (void)
+	{
```
Comment (hossamadel) [inline Dynamic Design/SERVICE/TMU/TMU.c]: Use space not tap

File: Dynamic Design/SERVICE/TMU/TMU.c [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+				/*Timer Configurations*/
+				Timer_Cfg_ST str_TimerCfg;
+				str_TimerCfg.Timer_Ch = (pstr_ConfigPtr->u8_Timer_Id);
+				str_TimerCfg.Timer_Mode = TIMER_MODE;
+				str_TimerCfg.Timer_PreScaler = PRESCALER_64;
+				str_TimerCfg.Timer_Interrupt_Mode = TIMER_MODE_INTERRUPT;
+				str_TimerCfg.Ptr_TCB_Function = Timer_CB;
+
+				Timer_Init(&str_TimerCfg);
+				/*store timer id in global variable to use in TMU_Start*/
```
Comment (hossamadel) [inline Dynamic Design/SERVICE/TMU/TMU.c]: Is the init timer return any error ? If yes please consider

File: Dynamic Design/SERVICE/TMU/TMU.c [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+		return Error;	
+	}
+
+/* this function is used to check timing for each task*/
+ ERROR_STATUS TMU_Dispatch(void)
+	{
+		uint8_t u8_Buffer_Index=0;
+		static uint8_t u8_Flag=0;
+		uint8_t Error = E_OK;
+		if (HIGH == gu8_Tick_Flag)
```
Comment (hossamadel) [inline Dynamic Design/SERVICE/TMU/TMU.c]: What I’m f we received more than one ISR before the dispatcher get invoked ?

File: Dynamic Design/SERVICE/TMU/TMU.c [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+		{
+			if(arrstr_TMU_Buffer[u8_Buffer_Index].u8_Duty == USED)
+			{
+				arrstr_TMU_Buffer[u8_Buffer_Index].u8_Pre_Flag++;
+
+				/*Check if this task will serve this tick*/
+				if  (arrstr_TMU_Buffer[u8_Buffer_Index].u8_Pre_Flag == (arrstr_TMU_Buffer[u8_Buffer_Index].u8_Delay_Time) )
+				{
+					
+					arrstr_TMU_Buffer[u8_Buffer_Index].Ptr_Consumer();
```
Comment (hossamadel) [inline Dynamic Design/SERVICE/TMU/TMU.c]: What if the consumer function equal NULL , you didn’t check it in the start function or here

File: Dynamic Design/SERVICE/TMU/TMU.c [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+						
+						
+
+
+			      }
+					//else for restart same task
+			      else if (au8_Already_Started == INITIALIZED)
+			      {
+				      aS8_Empty_Buffer_Location = EMPTY_BUFFER_LOCATION_INITIAL_VALUE;
+				      SET_BIT(PORTA_DATA,BIT6);
```
Comment (hossamadel) [inline Dynamic Design/SERVICE/TMU/TMU.c]: Put any debug line under #ifdef UNIT_TEST OR DEBUG

**Comment by hossamadel** [comment]
Thanks for your great work

File: Dynamic Design/SERVICE/TMU/TMU.c (hunk 7 above)
Comment (3dawy) [reply Dynamic Design/SERVICE/TMU/TMU.c]: start function checked it before stored in the buffer.

File: Dynamic Design/SERVICE/TMU/TMU.c (hunk 6 above)
Comment (3dawy) [reply Dynamic Design/SERVICE/TMU/TMU.c]: it is a design error from the user not from the module, is it??

**Pull request merged**